aten/src/ATen/functorch/BatchRulesScatterOps.cpp

```diff
@@ -1123,20 +1124,48 @@ std::tuple<Tensor,optional<int64_t>> index_fill_int_tensor_batch_rule_impl(
   auto batch_size = get_bdim_size3(self, self_bdim, index, index_bdim, value, value_bdim);
   self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
   index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
-  value_ = ensure_has_bdim(value_, value_bdim.has_value(), batch_size);
+
+  if (inplace || value_bdim.has_value()) {
+    // Do for-loop for in-place because we cannot reshape
+    // `self_` having an incompatible stride without copying.
+    // If value has a batch dim, we do for-loop as well because
+    // index_fill_ supports 1-element tensor only.
+    for (const auto i : c10::irange(0, batch_size)) {
+      const auto& self_slice = self_.select(0, i);
+      const auto& index_slice = index_.select(0, i);
+      self_slice.index_fill_(
+        dim,
+        index_slice,
+        value_bdim.has_value() ? value_.select(0, i) : value_
+      );
+    }
+    return std::make_tuple(self_, 0);
+  }
 
   self_ = self_bdim.has_value() ? self_ : self_.clone();
```

Comment:
IIUC, this line onwards is the same as `index_fill_int_scalar_batch_rule_impl` right? In that case, can we move this to helper function shared by both of them.
https://github.com/pytorch/pytorch/blob/beaaaaa9b4f8fdeddc21f5eef31fc3d786f3bad6/aten/src/ATen/functorch/BatchRulesScatterOps.cpp#L1067-L1093